### ISSUE TITLE
[v2022.05.103.27] M1Max macOS failed to run: System.PlatformNotSupportedException

### ISSUE THREAD

**Issue by sxysxy** [bluemagic123/tModLoader#2559]
### Version

Latest 1.4-stable

### OS

Mac

### Platform

Steam

### This bug affects

Gameplay as a Player

### Description

System&OS: M1Max, macOS12.4
Terraria source: Steam version 1.4.3.6
tModLoader source: Both Steam version(1.4 alpha) and  v2022.05.103.27(from GitHub) occurred this failure.
I just run ./start-tModLoader.sh 

### Log File

[01:24:43] [Main Thread/INFO] [tML]: Save Are Located At: /Users/hfcloud/Library/Application Support/Terraria/tModLoader
[01:24:43] [Main Thread/INFO] [tML]: Starting tModLoader client 1.4.3.6+2022.05.103.27|1.4-stable|Stable|a2eec54e5db5108348eea5278f3abc9205923bd7|5249591270184616505 built 2022/6/11 14:19
[01:24:43] [Main Thread/INFO] [tML]: Log date: 2022/6/16
[01:24:43] [Main Thread/INFO] [tML]: Running on OSX Arm64 NetCore 6.0.0
[01:24:43] [Main Thread/INFO] [tML]: Executable: /Users/hfcloud/Library/Application Support/Steam/steamapps/common/tModLoader/tModLoader.dll
[01:24:43] [Main Thread/INFO] [tML]: Working Directory: /Users/hfcloud/Library/Application Support/Steam/steamapps/common/tModLoader
[01:24:43] [Main Thread/WARN] [tML]: tModLoader.RuntimeErrorSilentlyCaughtException
System.PlatformNotSupportedException: Operation is not supported on this platform.
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform.GetJitObject()
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform.Create()
   at MonoMod.RuntimeDetour.DetourHelper.get_Runtime()
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig& config)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig config)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase method, Delegate to)
   at Terraria.ModLoader.Core.LoggingHooks.PrettifyStackTraceSources() in tModLoader\Terraria\ModLoader\Core\LoggingHooks.cs:line 38
   at Terraria.ModLoader.Logging.LogStartup(Boolean dedServ) in tModLoader\Terraria\ModLoader\Logging.cs:line 83
   at Terraria.Program.LaunchGame(String[] args, Boolean monoArgs) in tModLoader\Terraria\Program.cs:line 185
   at MonoLaunch.<>c__DisplayClass2_0.<Main>b__0() in tModLoader\Terraria\MonoLaunch.cs:line 65
   at MonoLaunch.Main(String[] args) in tModLoader\Terraria\MonoLaunch.cs:line 72

[01:24:43] [Main Thread/WARN] [tML]: tModLoader.RuntimeErrorSilentlyCaughtException
System.ComponentModel.Win32Exception: Permission denied
   at MonoMod.RuntimeDetour.Platforms.DetourNativeLibcPlatform.SetMemPerms(IntPtr start, UInt64 len, MmapProts prot)
   at MonoMod.RuntimeDetour.Platforms.DetourNativeLibcPlatform.MakeWritable(IntPtr src, UInt32 size)
   at MonoMod.RuntimeDetour.DetourHelper.MakeWritable(IDetourNativePlatform plat, NativeDetourData detour)
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeILPlatform._HookSelftest(MethodInfo from, MethodInfo to)
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeILPlatform..ctor()
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETPlatform..ctor()
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform..ctor()
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform.Create()
   at MonoMod.RuntimeDetour.DetourHelper.get_Runtime()
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig& config)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig config)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase method, Delegate to)
   at Terraria.ModLoader.Core.LoggingHooks.PrettifyStackTraceSources() in tModLoader\Terraria\ModLoader\Core\LoggingHooks.cs:line 38
   at Terraria.ModLoader.Logging.LogStartup(Boolean dedServ) in tModLoader\Terraria\ModLoader\Logging.cs:line 83
   at Terraria.Program.LaunchGame(String[] args, Boolean monoArgs) in tModLoader\Terraria\Program.cs:line 185
   at MonoLaunch.<>c__DisplayClass2_0.<Main>b__0() in tModLoader\Terraria\MonoLaunch.cs:line 65
   at MonoLaunch.Main(String[] args) in tModLoader\Terraria\MonoLaunch.cs:line 72

[01:24:43] [Main Thread/FATAL] [Terraria]: Failed to Handle Save Data
System.ComponentModel.Win32Exception (13): Permission denied
   at MonoMod.RuntimeDetour.Platforms.DetourNativeLibcPlatform.SetMemPerms(IntPtr start, UInt64 len, MmapProts prot)
   at MonoMod.RuntimeDetour.Platforms.DetourNativeLibcPlatform.MakeWritable(IntPtr src, UInt32 size)
   at MonoMod.RuntimeDetour.DetourHelper.MakeWritable(IDetourNativePlatform plat, NativeDetourData detour)
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeILPlatform._HookSelftest(MethodInfo from, MethodInfo to)
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeILPlatform..ctor()
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETPlatform..ctor()
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform..ctor()
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform.Create()
   at MonoMod.RuntimeDetour.DetourHelper.get_Runtime()
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig& config)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig config)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase method, Delegate to)
   at Terraria.ModLoader.Core.LoggingHooks.PrettifyStackTraceSources() in D:\a\tModLoader\tModLoader\src\tModLoader\Terraria\ModLoader\Core\LoggingHooks.cs:line 38
   at Terraria.ModLoader.Logging.LogStartup(Boolean dedServ) in D:\a\tModLoader\tModLoader\src\tModLoader\Terraria\ModLoader\Logging.cs:line 82
   at Terraria.Program.LaunchGame(String[] args, Boolean monoArgs) in D:\a\tModLoader\tModLoader\src\tModLoader\Terraria\Program.cs:line 185

### Steps to reproduce

Just run ./start-tModLoader.sh

**Comment by pascil** [comment]
How do I fix that? It's still present on the newest build, and no one seems to care on the discord.

**Comment by nike4613** [comment]
MonoMod doesn't currently support the M1, due to a combination of its funky API, memory protection requirements, and the current architecture of MonoMod. I have been working on a rewrite which should allow us to implement it more cleanly, though that still has some (fortunately relatively small) work before it can be sanely merged.

**Comment by morningtzh** [comment]
My intel macbook has this problem too. 

I had reinstall steam(and it's AppBundle) / terraria / tml(1.4), but not work.

when I install tml 1.3, it works. sadly my terraria savefile is 1.4. 

```
[21:53:25] [Main Thread/INFO] [tML]: Save Are Located At: /Users/morningtzh/Library/Application Support/Terraria/tModLoader
[21:53:25] [Main Thread/INFO] [tML]: Starting tModLoader client 1.4.3.6+2022.06.96.4|1.4-stable|Stable|6ee86304725fa4ac713d981d17a87d1b1d0e014a|5249609086868104401 built 7/2/2022 5:14 AM
[21:53:25] [Main Thread/INFO] [tML]: Log date: 7/15/2022
[21:53:25] [Main Thread/INFO] [tML]: Running on OSX X64 NetCore 6.0.0
[21:53:25] [Main Thread/INFO] [tML]: Executable: /Users/morningtzh/Library/Application Support/Steam/steamapps/common/tModLoader/tModLoader.dll
[21:53:25] [Main Thread/INFO] [tML]: Working Directory: /Users/morningtzh/Library/Application Support/Steam/steamapps/common/tModLoader
[21:53:25] [Main Thread/WARN] [tML]: tModLoader.RuntimeErrorSilentlyCaughtException
System.PlatformNotSupportedException: Operation is not supported on this platform.
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform.GetJitObject()
   at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform.Create()
   at MonoMod.RuntimeDetour.DetourHelper.get_Runtime()
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig& config)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig config)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase method, Delegate to)
   at Terraria.ModLoader.Core.LoggingHooks.PrettifyStackTraceSources() in tModLoader\Terraria\ModLoader\Core\LoggingHooks.cs:line 38
   at Terraria.ModLoader.Logging.LogStartup(Boolean dedServ) in tModLoader\Terraria\ModLoader\Logging.cs:line 83
   at Terraria.Program.LaunchGame(String[] args, Boolean monoArgs) in tModLoader\Terraria\Program.cs:line 186
   at MonoLaunch.<>c__DisplayClass2_0.<Main>b__0() in tModLoader\Terraria\MonoLaunch.cs:line 65
   at MonoLaunch.Main(String[] args) in tModLoader\Terraria\MonoLaunch.cs:line 72

[21:53:26] [Main Thread/INFO] [tML]: Distribution Platform: Steam. Detection method: CWD is /steamapps/
[21:53:26] [Main Thread/DEBUG] [tML]: Process.Start (UseShellExecute = False): "/Users/morningtzh/Library/Application Support/Steam/steamapps/common/tModLoader/dotnet/6.0.0/dotnet" tModLoader.dll -terrariasteamclient 230
[21:53:52] [Main Thread/DEBUG] [TerrariaSteamClient]: Recv: init_failed
[21:53:52] [Main Thread/DEBUG] [tML]: Process.Start (UseShellExecute = False): "open" "https://terraria.org"
[21:53:52] [Main Thread/FATAL] [tML]: Steam API hash mismatch, assumed pirated.

tModLoader requires a legitimate Terraria install to work.
```

**Comment by jhonathanwesley** [comment]
> My intel macbook has this problem too.
> 
> I had reinstall steam(and it's AppBundle) / terraria / tml(1.4), but not work.
> 
> when I install tml 1.3, it works. sadly my terraria savefile is 1.4.
> 
> ```
> [21:53:25] [Main Thread/INFO] [tML]: Save Are Located At: /Users/morningtzh/Library/Application Support/Terraria/tModLoader
> [21:53:25] [Main Thread/INFO] [tML]: Starting tModLoader client 1.4.3.6+2022.06.96.4|1.4-stable|Stable|6ee86304725fa4ac713d981d17a87d1b1d0e014a|5249609086868104401 built 7/2/2022 5:14 AM
> [21:53:25] [Main Thread/INFO] [tML]: Log date: 7/15/2022
> [21:53:25] [Main Thread/INFO] [tML]: Running on OSX X64 NetCore 6.0.0
> [21:53:25] [Main Thread/INFO] [tML]: Executable: /Users/morningtzh/Library/Application Support/Steam/steamapps/common/tModLoader/tModLoader.dll
> [21:53:25] [Main Thread/INFO] [tML]: Working Directory: /Users/morningtzh/Library/Application Support/Steam/steamapps/common/tModLoader
> [21:53:25] [Main Thread/WARN] [tML]: tModLoader.RuntimeErrorSilentlyCaughtException
> System.PlatformNotSupportedException: Operation is not supported on this platform.
>    at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform.GetJitObject()
>    at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform.Create()
>    at MonoMod.RuntimeDetour.DetourHelper.get_Runtime()
>    at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig& config)
>    at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig config)
>    at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target)
>    at MonoMod.RuntimeDetour.Hook..ctor(MethodBase method, Delegate to)
>    at Terraria.ModLoader.Core.LoggingHooks.PrettifyStackTraceSources() in tModLoader\Terraria\ModLoader\Core\LoggingHooks.cs:line 38
>    at Terraria.ModLoader.Logging.LogStartup(Boolean dedServ) in tModLoader\Terraria\ModLoader\Logging.cs:line 83
>    at Terraria.Program.LaunchGame(String[] args, Boolean monoArgs) in tModLoader\Terraria\Program.cs:line 186
>    at MonoLaunch.<>c__DisplayClass2_0.<Main>b__0() in tModLoader\Terraria\MonoLaunch.cs:line 65
>    at MonoLaunch.Main(String[] args) in tModLoader\Terraria\MonoLaunch.cs:line 72
> 
> [21:53:26] [Main Thread/INFO] [tML]: Distribution Platform: Steam. Detection method: CWD is /steamapps/
> [21:53:26] [Main Thread/DEBUG] [tML]: Process.Start (UseShellExecute = False): "/Users/morningtzh/Library/Application Support/Steam/steamapps/common/tModLoader/dotnet/6.0.0/dotnet" tModLoader.dll -terrariasteamclient 230
> [21:53:52] [Main Thread/DEBUG] [TerrariaSteamClient]: Recv: init_failed
> [21:53:52] [Main Thread/DEBUG] [tML]: Process.Start (UseShellExecute = False): "open" "https://terraria.org"
> [21:53:52] [Main Thread/FATAL] [tML]: Steam API hash mismatch, assumed pirated.
> 
> tModLoader requires a legitimate Terraria install to work.
> ```

This same error ocurred with me: 

![imagem_2022-07-15_113147501](https://user-images.githubusercontent.com/98562102/179244837-204e93a2-c537-43eb-a0b6-41eb4f7deeb4.png)

And I installed the 1.3 beta and it worked well...

**Comment by morningtzh** [comment]
> My intel macbook has this problem too.
> 
> I had reinstall steam(and it's AppBundle) / terraria / tml(1.4), but not work.
> 
> when I install tml 1.3, it works. sadly my terraria savefile is 1.4.
> 
> ```
> [21:53:25] [Main Thread/INFO] [tML]: Save Are Located At: /Users/morningtzh/Library/Application Support/Terraria/tModLoader
> [21:53:25] [Main Thread/INFO] [tML]: Starting tModLoader client 1.4.3.6+2022.06.96.4|1.4-stable|Stable|6ee86304725fa4ac713d981d17a87d1b1d0e014a|5249609086868104401 built 7/2/2022 5:14 AM
> [21:53:25] [Main Thread/INFO] [tML]: Log date: 7/15/2022
> [21:53:25] [Main Thread/INFO] [tML]: Running on OSX X64 NetCore 6.0.0
> [21:53:25] [Main Thread/INFO] [tML]: Executable: /Users/morningtzh/Library/Application Support/Steam/steamapps/common/tModLoader/tModLoader.dll
> [21:53:25] [Main Thread/INFO] [tML]: Working Directory: /Users/morningtzh/Library/Application Support/Steam/steamapps/common/tModLoader
> [21:53:25] [Main Thread/WARN] [tML]: tModLoader.RuntimeErrorSilentlyCaughtException
> System.PlatformNotSupportedException: Operation is not supported on this platform.
>    at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform.GetJitObject()
>    at MonoMod.RuntimeDetour.Platforms.DetourRuntimeNETCorePlatform.Create()
>    at MonoMod.RuntimeDetour.DetourHelper.get_Runtime()
>    at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig& config)
>    at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target, HookConfig config)
>    at MonoMod.RuntimeDetour.Hook..ctor(MethodBase from, MethodInfo to, Object target)
>    at MonoMod.RuntimeDetour.Hook..ctor(MethodBase method, Delegate to)
>    at Terraria.ModLoader.Core.LoggingHooks.PrettifyStackTraceSources() in tModLoader\Terraria\ModLoader\Core\LoggingHooks.cs:line 38
>    at Terraria.ModLoader.Logging.LogStartup(Boolean dedServ) in tModLoader\Terraria\ModLoader\Logging.cs:line 83
>    at Terraria.Program.LaunchGame(String[] args, Boolean monoArgs) in tModLoader\Terraria\Program.cs:line 186
>    at MonoLaunch.<>c__DisplayClass2_0.<Main>b__0() in tModLoader\Terraria\MonoLaunch.cs:line 65
>    at MonoLaunch.Main(String[] args) in tModLoader\Terraria\MonoLaunch.cs:line 72
> 
> [21:53:26] [Main Thread/INFO] [tML]: Distribution Platform: Steam. Detection method: CWD is /steamapps/
> [21:53:26] [Main Thread/DEBUG] [tML]: Process.Start (UseShellExecute = False): "/Users/morningtzh/Library/Application Support/Steam/steamapps/common/tModLoader/dotnet/6.0.0/dotnet" tModLoader.dll -terrariasteamclient 230
> [21:53:52] [Main Thread/DEBUG] [TerrariaSteamClient]: Recv: init_failed
> [21:53:52] [Main Thread/DEBUG] [tML]: Process.Start (UseShellExecute = False): "open" "https://terraria.org"
> [21:53:52] [Main Thread/FATAL] [tML]: Steam API hash mismatch, assumed pirated.
> 
> tModLoader requires a legitimate Terraria install to work.
> ```

I deleted all dotnet in /usr/local/share and the dotnet in tmodloader folder. It works.

tmodloader can't find any dotnet, will redownload from microsoft. 

And it not work when I only remove tmodloader in steam, the dotnet in tmodloader will remain. After I only reinstall tmodloader, modified time of the dotnet folder  is last week.

> [How to remove the .NET Runtime and SDK](https://docs.microsoft.com/en-us/dotnet/core/install/remove-runtime-sdk-versions?pivots=os-macos)

**Comment by JavidPack** [comment]
Is this still an issue? We aren't sure if this has been fixed with recent changes or not.

**Comment by nike4613** [comment]
No changes have been made on the MonoMod side to support M1 yet, due to underlying architectural issues which make supporting the M1 quite difficult. A major rewrite is still in-progress to address this and other issues (and I believe there is a tML branch using it) but as I do not have access to an M1, it does not support it either at the moment.

**Comment by tmuaprep** [comment]
any update?

**Comment by Chicken-Bones** [comment]
Fixed in 1.4.4
